test/k8sT/CustomCalls.go

```diff
@@ -351,5 +355,83 @@ var _ = Describe("K8sCustomCalls", func() {
 
 			checkByteCounter(options, expectedByteCount, 0, false)
 		})
+
+		getMissedCustomCallsCount := func(ciliumPod string,
+			direction string) int {
+
+			cmd := fmt.Sprintf("cilium bpf metrics list -o jsonpath='{$[?(@.reason==11)].values.%s.packets}'", direction)
+			res := kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			res.ExpectSuccess("Failed to lookup metrics for missed tail calls to custom programs")
+
+			// If the metrics is missing from the output, consider
+			// it is a zero value
+			output := strings.TrimSpace(res.Stdout())
+			if output == "" {
+				return 0
+			}
+
+			count, err := strconv.Atoi(output)
+			ExpectWithOffset(2, err).ToNot(HaveOccurred(),
+				fmt.Sprintf("Failed to convert metrics value: %s", err))
+			return count
+		}
+
+		checkMetrics := func(enableCustomCalls bool) {
+			var customCallsOption string
+
+			if enableCustomCalls {
+				customCallsOption = "true"
+			} else {
+				customCallsOption = "false"
+			}
+
+			deploymentManager.DeployCilium(map[string]string{
+				"customCalls.enabled": customCallsOption,
+			}, DeployCiliumOptionsAndDNS)
+
+			ciliumPodK8s2, err := kubectl.GetCiliumPodOnNode(helpers.K8s2)
+			Expect(err).ShouldNot(HaveOccurred(), "Cannot get cilium pod on k8s2")
+
+			getPodsInfo()
+
+			err = kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=testapp", helpers.HelperTimeout)
+			ExpectWithOffset(1, err).Should(BeNil())
+
+			// Send traffic between pods
+			cmd := helpers.PingWithCount(podApp1.Status.PodIP, 1)
+			res := kubectl.ExecPodCmd(helpers.DefaultNamespace, podApp2.Name, cmd)
+			res.ExpectSuccess(fmt.Sprintf("Failed to ping from %s to %s", podApp2.Name, podApp1.Status.PodIP))
+
+			// Check metrics
+			for _, direction := range []string{"ingress", "egress"} {
+				missedCalls := getMissedCustomCallsCount(ciliumPodK8s2, direction)
+				if enableCustomCalls {
+					ExpectWithOffset(1, missedCalls).To(BeNumerically(">", 0),
+						fmt.Sprintf("Zero value for metrics (%s) when tail calls to custom programs are enabled", direction))
+				} else {
+					ExpectWithOffset(1, missedCalls).To(Equal(0),
+						fmt.Sprintf("Non-zero value for metrics (%s) when tail calls to custom programs are disabled", direction))
+				}
+			}
+		}
+
+		It("Increments dedicated metrics on missed tail calls", func() {
+
+			By("Resetting metrics")
+
+			cmd := "./cilium/cilium cleanup --force"
+			res := kubectl.ExecPodCmd(helpers.DefaultNamespace, compilerPodName, cmd)
+			res.ExpectSuccess("Failed to clean Cilium state (and metrics)")
+
+			By("Checking metrics are at 0 when tail calls to custom programs are disabled")
+
+			checkMetrics(false)
+
+			deploymentManager.DeleteAll()
+
+			By("Checking metrics are non-nul when tail calls to custom programs are enabled")
+
+			checkMetrics(true)
+		})
```

Comment:
Why can't this be part of the existing tests?
1. Deploy Cilium with custom calls.
2. Check the drop metric increases.
3. Load the custom programs.
4. Check the drop metric stopped increasing.

Doing so avoids the need for `cilium cleanup`. I think it's fine if we don't check the metric when Cilium isn't compiled with custom calls; that case is pretty straightforward.

It would also allow us to check with and without endpoint routes.

Reply:
I can do that, but we'll have the following limitations:

- [We won't be able to tell that the metrics remain at zero when custom calls are disabled, but as you say this is probably not a big concern. Maybe in the future we could try attaching a program to _all_ available hooks in the cluster and check that the metrics remain stable.]
- If another test ever enables the calls to custom programs before the byte counter tests run, we won't be able to tell that the metrics is being increased for the byte counter tests.
- The test with per-endpoint routes will bring no useful info, because the metrics will already have a non-zero value from the previous test. Maybe I can collect the metrics twice to check that it does change (even if we'll have no info on what hooks make the metrics raise, but this is beyond the scope of the test probably). This could solve the previous item too, actually, I might go this way.

Reply:
> If another test ever enables the calls to custom programs before the byte counter tests run, we won't be able to tell that the metrics is being increased for the byte counter tests.

Why? By `Check the drop metric` I meant take it twice with a couple seconds sleep in between and compare.

Reply:
:+1: I hadn't understood at first, but that's what I said I'll do in my third item above :)

Reply:
So you mean I could have read only the last sentence of your previous answer?

Reply:
Well you wouldn't have known what would have solved “the previous item” or what way I would go if you had read _only_ the last _sentence_, I guess :)